keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,879 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### OpenAPI v3 is a richer API description than OpenAPI v2
+
+OpenAPI v3 support in Kubernetes is currently beta since version 1.24.
+OpenAPI V3 is a richer representation of the kubernetes API to our users, who have been asking for visibility
+into things like:
+
+1. nullable
+2. default
+3. validation fields like oneOf, anyOf, etc.
+
+To show each of these additional data points by themselves is a strong reason
+to switch to using OpenAPI v3.
+
+
+### CRD schemas expressed as OpenAPI v2 are lossy
+
+Today CRDs specify their schemas in OpenAPI v3 format. To serve the `/openapi/v2`
+document used today by kubectl, there is an expensive conversion from the v3 down
+to v2 format.
+
+This process is [very lossy](https://github.com/kubernetes/kubernetes/blob/6e0de20fbb4c127d2e45c7a22347c08545fc7a86/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go#L56-L66), so `kubectl explain` when used against CRDs
+making use of v3 features does not have a good experience with inaccurate information, or fields removed altogther.
+
+This transformation causes bugs, for example, when attempting to `explain` a field
+that is `nullable`, kubectl instead shows nothing, due to the lossy conversion
+wiping nullable fields.
+
+### Goals
+
+1. Provide the new richer type information specified by OpenAPI v3 within kubectl explain
+2. Have a more maintainable `text/template` based approach to printing
+3. Fallback to old `explain` implementation if cluster does not expose OpenAPI v3 data.
+4. Provide multiple new output formats for kubectl explain:
+    * human-readable plaintext
+    * markdown
+    * maybe others
+5. (Optional?) Allow users to specify their own templates for use with kubectl
+  explain (there may be interesting use cases for this)
+
+### Non-Goals
+
+1. "Fix" openapi v3 to openapi v2 conversion
+  This is a non-goal for two reasons:
+    * These formats are not compatible, and there WILL be data loss and inaccuracy
+    * This negates the benefits of using OpenAPI v3 for the richer type information
+2. Provide general-purpose OpenAPI visualization.
+
+
+## Proposal
+
+
+### Basic Usage
+
+The following user experience should be possible with `kubectl explain`
+
+```shell
+kubectl explain pods
+```
+
+Output should be familiar to users of today's `kubectl explain`, except new
+information is populated:
+
+```console
+<TODO>
+```
+
+### Built-in Template Options
+
+Users should be able to specify built-in templates to use instead of the 
+human-readable plaintext form:
+```shell
+kubectl explain pods --template md
+```
+
+When using the `md` template, a markdown document is printed to stdout, so it
+might be saved and used for a documentation website, for example.
+
+```console
+<TODO>
+```
+
+### Custom Templates
```

Comment:
Do you have some examples of users requesting this? On the one hand, we have precedent for template support in  `kubectl get`: `--output=go-template-file --template=path/to/file`. On the other, in those cases, the end user is exposed to the schema of the resource they're requesting and are able to control it (e.g. `kubectl get deployment.v1.apps`), whereas they don't see/control the version of the openapi data this template would ingest.

If we do find the use case for custom templates compelling, perhaps we need to expose that choice explicitly so that we have a deprecation path if ever there's an openapi v4 we want to adopt. That could be the same flag we use to feature-gate this, e.g. `--openapi-version=v3-experimental` or something.

An alternative stance could be that openapi v3 isn't kubernetes-specific, and the use cases for consuming it could go beyond templating. So instead of template options, we could have an output option to expose it raw, i.e. `--output=openapiv3`, and expect tools (like kubectl extensions) to build display functionality around that.

Reply:
Internally we have some use cases, but not for custom templates since the proposed builtin formats are sufficient. Since there is no concrete use case and to reduce bloating scope, I've cut custom templates out from the KEP.  

Once we have use cases, and know how we want to use the data through the built-in templates, we can consider custom templates in the future.